web/react/components/modal.jsx

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2015 Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+export default class Modal extends ReactBootstrap.Modal {
```

Comment:
I agree that we'll need to be careful about how we use onShow, but it's no different than what we can currently do in componentDidUpdate or other lifecycle methods.

Some of the examples that I gave aren't the best (we still need to come up with a better way to set the max height of modals), but there's still cases where it's valuable. Looking at the AccessHistoryModal, it requests stuff from the server when its shown since we don't need that data pulled otherwise and I think that's a good use case for this. There's also occasionally setup of UI state like clearing error messages which doesn't need to go back to the full stores.

In those cases, onShow is roughly equivalent to componentDidMount for these modal components. React-Bootstrap modals are mounted as soon as their parents are, even if they're not shown until much later. Having onShow lets us delay stuff until it's actually needed and allows it to trigger again if the dialog is hidden and then shown a second time.

Also, I looked into why they haven't included it in React-Bootstrap and they do have onEntering/onEntered which work like onPreshow/onShow, but they require that the modal is animated to work. There was some discussion about onShow, but they wanted something that was more similar to the existing onHide. They did say that a custom [ReactBootstrap.ModalDialog](https://github.com/react-bootstrap/react-bootstrap/blob/master/src/ModalDialog.js) could have proper lifecycle handles attached to it, but that seems like it's unnecessary for what we're doing since it would necessitate having two components for each modal.

For reference, I think using a custom ModalDialog would look like this:
```javascript
class StuffModal extends React.Component {
    render() {
        return (
            <Modal show={this.props.show} dialogComponent={StuffModalDialog}>
                <Contents />
            </Modal>
        );
    }
}

class StuffModalDialog extends ReactBootstrap.ModalDialog {
    componentDidMount() {
        doStuff();
    }
}
```
Alternatively, Contents could be owned by StuffModalDialog, but that would require re-implementing the render method of ModalDialog.